fontFeatures/fontGlyphs.py

```diff
@@ -0,0 +1,184 @@
+        # https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#table-organization
+
+        lookups = []
+        # assign DFLT to script if selected script does not exist
+        # assign dflt to lang if selected lang does not exist
+        if script not in self.unparsed.languageSystems:
+            script = "DFLT"
+        if lang not in self.unparsed.languageSystems[script]:
+            lang = "dflt"
+
+        for feat in self.unparsed.features:
+            if feat not in selected_features:
+                continue
+            if script not in self.unparsed.features[feat]:
+                continue
+            if lang not in self.unparsed.features[feat][script]:
+                continue
+            lookups += [(feat, self.unparsed.features[feat][script][lang])]
+
+        # Sort the lookups by the order given in the LookupList table.
+        res = []
+        for feat, lk_idxs in sorted(lookups, key=lambda k: k[1]):
+            for idx in lk_idxs:
+                res.append((feat, self.unparsed.lookups[idx]))
+        return res
+
+    def _process_routine(self, tag, routine):
+        for rule in routine.rules:
+            if routine.name.startswith("SingleSubstitution"):
+                self._process_simple_subs(tag, rule)
+            elif routine.name.startswith("LigatureSubstitution"):
+                self._process_simple_subs(tag, rule)
+            elif routine.name.startswith("AlternateSubstitution"):
+                self._process_alt_subs(tag, rule)
+            elif routine.name.startswith("ChainedContextual"):
+                self._process_chained_subs(tag, rule)
+            else:
+                print(f"Cannot process {rule.name}")
+
+    def _memoize_input(self, input_, tag):
+        res = {"input": "", "features": set([tag])}
+        for glyph in input_:
+            if glyph in self.glyph_inputs:
+                res["features"] |= self.glyph_inputs[glyph]["features"]
+                res["input"] += self.glyph_inputs[glyph]["input"]
+        return res
+
+    def _process_simple_subs(self, tag, rule):
+        ## [["A"], ["B"], ["C", "D"]] --> [('A', 'B', 'C'), ('A', 'B', 'D')]
+        inputs = list(itertools.product(*rule.input))
+        for input_, result in zip(inputs, rule.replacement[0]):
+            if result in self.glyph_inputs:
+                continue
+            self.glyph_inputs[result] = self._memoize_input(input_, tag)
+
+    def _process_alt_subs(self, tag, rule):
+        inputs = list(itertools.product(*rule.replacement))
+        for input_, result in zip(inputs, rule.input[0]):
+            if result in self.glyph_inputs:
+                continue
+            self.glyphs_input[result] = self._memoize_input(input_, tag)
+
+    def _process_chained_subs(self, tag, rule):
+        inputs = list(
+            itertools.product(*rule.precontext, *rule.input, *rule.postcontext)
+        )
+        # process lks
+        for lk in rule.lookups[0]:
```

Comment:
When I run this on a production font, I have cycles in this for loop where `rule.lookups[0]` is None.  This leads to an exception.

Reply:
Done some digging.

For the frac feature `itertools.product(*rule.precontext, *rule.input, *rule.postcontext)` is producing more glyph combos than people exist on earth 🤣  (10^10).

Reply:
> For the frac feature `itertools.product(*rule.precontext, *rule.input, *rule.postcontext)` is producing more glyph combos than people exist on earth 🤣 (10^10).

Indeed, you should return sequence of sets, instead of expanding them...